contrib/ruby/ext/trilogy-ruby/cext.c

```diff
@@ -28,26 +38,59 @@ static ID id_socket, id_host, id_port, id_username, id_password, id_found_rows,
     id_from_code, id_from_errno, id_connection_options, id_max_allowed_packet;
 
 struct trilogy_ctx {
+    VALUE self;
+
     trilogy_conn_t conn;
+
+#ifdef TRILOGY_RB_IO_WAIT
+    VALUE io;
+#endif
+
     char server_version[TRILOGY_SERVER_VERSION_SIZE + 1];
     unsigned int query_flags;
     VALUE encoding;
 };
 
-static void mark_trilogy(void *ptr)
+static void trilogy_ctx_mark(void *ptr)
 {
     struct trilogy_ctx *ctx = ptr;
-    rb_gc_mark(ctx->encoding);
+
+    rb_gc_mark_movable(ctx->self);
+
+#ifdef TRILOGY_RB_IO_WAIT
+    if (RTEST(ctx->io))
+        rb_gc_mark_movable(ctx->io);
```

Comment:
```suggestion
    rb_gc_mark_movable(ctx->io);
```

`gc_mark` don't mind `false/nil`

Reply:
Thanks I wonder where I got the idea I should check first.